# Model4_standardize

The purpose of this notebook / model experiment is to improve upon the previous fix_imbalance model achieved from model2_fix_imbalance.ipynb, by standardizing variables (normalizing + transforming)

In [1]:
import pandas as pd
from pycaret.classification import *

## import data

In [2]:
df_ts_agg = pd.read_csv("../proData/df_ts_agg.csv")
df_ts_agg = df_ts_agg.set_index("PATIENT_ID")

df_ts_agg

ihd  Age  Gender  Height                        ICUType  \
PATIENT_ID                                                            
132539        0   54  female     NaN                   Surgical ICU   
132540        0   76    male   175.3  Cardiac Surgery Recovery Unit   
132541        0   44  female     NaN                    Medical ICU   
132543        0   68    male   180.3                    Medical ICU   
132545        0   88  female     NaN                    Medical ICU   
...         ...  ...     ...     ...                            ...   
142665        0   70  female     NaN                   Surgical ICU   
142667        0   25    male     NaN                    Medical ICU   
142670        0   44    male     NaN                    Medical ICU   
142671        1   37    male     NaN                    Medical ICU   
142673        0   78  female   157.5                   Surgical ICU   

            Weight_min  Weight_max  Weight_med  GCS_min  GCS_max  ...  \
PATIENT_ID                                                        ...   
132539             NaN         NaN         NaN     14.0     15.0  ...   
132540            76.0        81.6        80.6      3.0     15.0  ...   
132541            56.7        56.7        56.7      5.0      8.0  ...   
132543            84.6        84.6        84.6     14.0     15.0  ...   
132545             NaN         NaN         NaN     15.0     15.0  ...   
...                ...         ...         ...      ...      ...  ...   
142665            87.0        87.0        87.0      3.0     15.0  ...   
142667           166.4       166.4       166.4     15.0     15.0  ...   
142670           109.0       109.0       109.0      3.0      8.0  ...   
142671            87.4        87.4        87.4      3.0      7.0  ...   
142673            70.7        88.8        88.8      3.0     15.0  ...   

            Lactate_med  Cholesterol_min  Cholesterol_max  Cholesterol_med  \
PATIENT_ID                                                                   
132539              NaN              NaN              NaN              NaN   
132540              NaN              NaN              NaN              NaN   
132541              1.3              NaN              NaN              NaN   
132543              NaN              NaN              NaN              NaN   
132545              NaN              NaN              NaN              NaN   
...                 ...              ...              ...              ...   
142665              2.3              NaN              NaN              NaN   
142667              NaN            117.0            117.0            117.0   
142670              NaN              NaN              NaN              NaN   
142671              1.9              NaN              NaN              NaN   
142673              2.6              NaN              NaN              NaN   

            TroponinI_min  TroponinI_max  TroponinI_med  TroponinT_min  \
PATIENT_ID                                                               
132539                NaN            NaN            NaN            NaN   
132540                NaN            NaN            NaN            NaN   
132541                NaN            NaN            NaN            NaN   
132543                NaN            NaN            NaN            NaN   
132545                NaN            NaN            NaN            NaN   
...                   ...            ...            ...            ...   
142665                NaN            NaN            NaN            NaN   
142667                NaN            NaN            NaN            NaN   
142670                NaN            NaN            NaN            NaN   
142671                NaN            NaN            NaN            NaN   
142673                NaN            NaN            NaN            NaN   

            TroponinT_max  TroponinT_med  
PATIENT_ID                                
132539                NaN            NaN  
132540                NaN      

## setup experiment with pycaret

with "normalize" and "transformation" set as true, some linear kernal-based model performance is expected to be improved

In [3]:
exp_physionet = setup(data = df_ts_agg, target = 'ihd', session_id=123,
                      numeric_features = ["GCS_min", "GCS_max"],
                      normalize = True, 
                      transformation = True, 
                      #ignore_low_variance = True,
                      #remove_multicollinearity = True, multicollinearity_threshold = 0.9,
                      fix_imbalance = True
                     ) 

Setup Succesfully Completed!


## compare common classification models

min(Recall, Precision) is used for comparing, as suggested by the original physionet 2012 challenge

In [9]:
compare_models(turbo = False)

ExtraTreesClassifier(bootstrap=False, ccp_alpha=0.0, class_weight=None,
                     criterion='gini', max_depth=None, max_features='auto',
                     max_leaf_nodes=None, max_samples=None,
                     min_impurity_decrease=0.0, min_impurity_split=None,
                     min_samples_leaf=1, min_samples_split=2,
                     min_weight_fraction_leaf=0.0, n_estimators=100, n_jobs=-1,
                     oob_score=False, random_state=123, verbose=0,
                     warm_start=False)

## create, tune, evaluate, predict gradient boosting classifier

In [4]:
gbc = create_model("gbc")

In [9]:
tuned_gbc = tune_model(gbc, optimize = "F1", choose_better = True)

In [10]:
evaluate_model(tuned_gbc)

interactive(children=(ToggleButtons(description='Plot Type:', icons=('',), options=(('Hyperparameters', 'param…

In [11]:
predict_model(tuned_gbc)

Model  Accuracy     AUC  Recall  Prec.      F1  \
0  Gradient Boosting Classifier    0.8485  0.8154  0.3675  0.442  0.4013   

    Kappa     MCC  
0  0.3154  0.3172

Age    Height  Weight_min  Weight_max  Weight_med   GCS_min  \
0     0.477499 -1.215611    0.547075    0.925603    0.944387 -1.152209   
1    -0.556735 -0.031642   -0.550466   -0.694846   -0.656055  1.181913   
2     1.769352  0.035305   -0.619652   -0.633653   -0.724077  0.669598   
3    -1.189793 -0.031642    0.179759    0.178626    0.175316  1.341102   
4    -0.057864 -0.031642   -2.213802   -2.314832   -2.271104  0.288248   
...        ...       ...         ...         ...         ...       ...   
1196  1.009600 -1.509279   -0.625997   -0.725754   -0.686850  0.483924   
1197 -0.505915 -0.031642   -0.361671   -0.478512   -0.440575  1.181913   
1198  0.406102 -0.031642   -0.677094   -0.819713   -0.780473  1.181913   
1199 -2.067770 -0.031642    0.575774    0.439906    0.472466 -0.142452   
1200  1.089991 -2.097840   -0.984866   -0.329244   -0.291976 -1.152209   

       GCS_max   GCS_med    HR_min    HR_max  ...  ICUType_Medical ICU  \
0     0.672376 -0.511354  0.753785 -0.510431  ...                  0.0   
1     0.672376  0.963869  1.427515  1.920192  ...                  1.0   
2     0.672376  0.434389 -1.360105  0.149794  ...                  0.0   
3     0.672376  0.963869  0.616506  0.103027  ...                  0.0   
4    -1.528718 -0.862434 -1.360105 -1.641198  ...                  0.0   
...        ...       ...       ...       ...  ...                  ...   
1196 -0.551305 -0.364897  0.055875 -0.295979  ...                  0.0   
1197  0.672376  0.963869 -0.827196  0.704247  ...                  0.0   
1198  0.672376  0.963869  2.408744  1.190597  ...                  1.0   
1199 -0.551305 -1.206565  1.025611  0.501095  ...                  0.0   
1200  0.672376 -0.640920  0.822069 -0.906773  ...                  0.0   

      ICUType_Surgical ICU  MechVent_min_0  MechVent_max_0  MechVent_max_1  \
0                      0.0             1.0             0.0             1.0   
1                      0.0             1.0             1.0             0.0   
2                      1.0             1.0             1.0             0.0   
3                      1.0             1.0             1.0             0.0   
4                      0.0             1.0             0.0             1.0   
...                    ...             ...             ...             ...   
1196                   1.0             1.0             1.0             0.0   
1197                   0.0             1.0             1.0             0.0   
1198                   0.0             1.0             1.0             0.0   
1199                   1.0             1.0             0.0             1.0   
1200                   0.0             1.0             0.0             1.0   

      MechVent_med_0.0  MechVent_med_0.5  ihd  Label   Score  
0                  1.0               0.0    0      0  0.0605  
1                  1.0               0.0    0      0  0.3409  
2                  1.0               0.0    1      0  0.3330  
3                  1.0               0.0    0      0  0.0219  
4                  1.0               0.0    0      0  0.1116  
...                ...               ...  ...    ...     ...  
1196               1.0               0.0    1      0  0.4648  
1197               1.0               0.0    0      0  0.0260  
1198               1.0               0.0    1      0  0.2266  
1199               1.0               0.0    0      0  0.1668  
1200               1.0               0.0    0      0  0.1486  

[1201 rows x 125 columns]

## explore blend models, both hard and soft voting

In [12]:
blend_hard = blend_models()

In [13]:
evaluate_model(blend_hard)

interactive(children=(ToggleButtons(description='Plot Type:', icons=('',), options=(('Hyperparameters', 'param…

In [14]:
predict_model(blend_hard)

Model  Accuracy  AUC  Recall   Prec.      F1   Kappa     MCC
0  Voting Classifier    0.8393    0  0.4337  0.4211  0.4273  0.3339  0.3339

Age    Height  Weight_min  Weight_max  Weight_med   GCS_min  \
0     0.477499 -1.215611    0.547075    0.925603    0.944387 -1.152209   
1    -0.556735 -0.031642   -0.550466   -0.694846   -0.656055  1.181913   
2     1.769352  0.035305   -0.619652   -0.633653   -0.724077  0.669598   
3    -1.189793 -0.031642    0.179759    0.178626    0.175316  1.341102   
4    -0.057864 -0.031642   -2.213802   -2.314832   -2.271104  0.288248   
...        ...       ...         ...         ...         ...       ...   
1196  1.009600 -1.509279   -0.625997   -0.725754   -0.686850  0.483924   
1197 -0.505915 -0.031642   -0.361671   -0.478512   -0.440575  1.181913   
1198  0.406102 -0.031642   -0.677094   -0.819713   -0.780473  1.181913   
1199 -2.067770 -0.031642    0.575774    0.439906    0.472466 -0.142452   
1200  1.089991 -2.097840   -0.984866   -0.329244   -0.291976 -1.152209   

       GCS_max   GCS_med    HR_min    HR_max  ...  ICUType_Coronary Care Unit  \
0     0.672376 -0.511354  0.753785 -0.510431  ...                         0.0   
1     0.672376  0.963869  1.427515  1.920192  ...                         0.0   
2     0.672376  0.434389 -1.360105  0.149794  ...                         0.0   
3     0.672376  0.963869  0.616506  0.103027  ...                         0.0   
4    -1.528718 -0.862434 -1.360105 -1.641198  ...                         0.0   
...        ...       ...       ...       ...  ...                         ...   
1196 -0.551305 -0.364897  0.055875 -0.295979  ...                         0.0   
1197  0.672376  0.963869 -0.827196  0.704247  ...                         0.0   
1198  0.672376  0.963869  2.408744  1.190597  ...                         0.0   
1199 -0.551305 -1.206565  1.025611  0.501095  ...                         0.0   
1200  0.672376 -0.640920  0.822069 -0.906773  ...                         0.0   

      ICUType_Medical ICU  ICUType_Surgical ICU  MechVent_min_0  \
0                     0.0                   0.0             1.0   
1                     1.0                   0.0             1.0   
2                     0.0                   1.0             1.0   
3                     0.0                   1.0             1.0   
4                     0.0                   0.0             1.0   
...                   ...                   ...             ...   
1196                  0.0                   1.0             1.0   
1197                  0.0                   0.0             1.0   
1198                  1.0                   0.0             1.0   
1199                  0.0                   1.0             1.0   
1200                  0.0                   0.0             1.0   

      MechVent_max_0  MechVent_max_1  MechVent_med_0.0  MechVent_med_0.5  ihd  \
0                0.0             1.0               1.0               0.0    0   
1                1.0             0.0               1.0               0.0    0   
2                1.0             0.0               1.0               0.0    1   
3                1.0             0.0               1.0               0.0    0   
4                0.0             1.0               1.0               0.0    0   
...              ...             ...               ...               ...  ...   
1196             1.0             0.0               1.0               0.0    1   
1197             1.0             0.0               1.0               0.0    0   
1198             1.0             0.0               1.0               0.0    1   
1199             0.0             1.0               1.0               0.0    0   
1200             0.0             1.0               1.0               0.0    0   

      Label  
0         0  
1         0  
2         0  
3         0  
4         0  
...     ...  
1196      1  
1197      0  
1198      0  
1199      0  
1200      0  

[1201 rows x 124 columns]

In [15]:
blend_soft = blend_models(method = "soft")

In [16]:
evaluate_model(blend_soft)

interactive(children=(ToggleButtons(description='Plot Type:', icons=('',), options=(('Hyperparameters', 'param…

In [17]:
predict_model(blend_soft)

Model  Accuracy    AUC  Recall   Prec.      F1   Kappa     MCC
0  Voting Classifier    0.8285  0.818  0.4157  0.3876  0.4012  0.3012  0.3015

Age    Height  Weight_min  Weight_max  Weight_med   GCS_min  \
0     0.477499 -1.215611    0.547075    0.925603    0.944387 -1.152209   
1    -0.556735 -0.031642   -0.550466   -0.694846   -0.656055  1.181913   
2     1.769352  0.035305   -0.619652   -0.633653   -0.724077  0.669598   
3    -1.189793 -0.031642    0.179759    0.178626    0.175316  1.341102   
4    -0.057864 -0.031642   -2.213802   -2.314832   -2.271104  0.288248   
...        ...       ...         ...         ...         ...       ...   
1196  1.009600 -1.509279   -0.625997   -0.725754   -0.686850  0.483924   
1197 -0.505915 -0.031642   -0.361671   -0.478512   -0.440575  1.181913   
1198  0.406102 -0.031642   -0.677094   -0.819713   -0.780473  1.181913   
1199 -2.067770 -0.031642    0.575774    0.439906    0.472466 -0.142452   
1200  1.089991 -2.097840   -0.984866   -0.329244   -0.291976 -1.152209   

       GCS_max   GCS_med    HR_min    HR_max  ...  ICUType_Medical ICU  \
0     0.672376 -0.511354  0.753785 -0.510431  ...                  0.0   
1     0.672376  0.963869  1.427515  1.920192  ...                  1.0   
2     0.672376  0.434389 -1.360105  0.149794  ...                  0.0   
3     0.672376  0.963869  0.616506  0.103027  ...                  0.0   
4    -1.528718 -0.862434 -1.360105 -1.641198  ...                  0.0   
...        ...       ...       ...       ...  ...                  ...   
1196 -0.551305 -0.364897  0.055875 -0.295979  ...                  0.0   
1197  0.672376  0.963869 -0.827196  0.704247  ...                  0.0   
1198  0.672376  0.963869  2.408744  1.190597  ...                  1.0   
1199 -0.551305 -1.206565  1.025611  0.501095  ...                  0.0   
1200  0.672376 -0.640920  0.822069 -0.906773  ...                  0.0   

      ICUType_Surgical ICU  MechVent_min_0  MechVent_max_0  MechVent_max_1  \
0                      0.0             1.0             0.0             1.0   
1                      0.0             1.0             1.0             0.0   
2                      1.0             1.0             1.0             0.0   
3                      1.0             1.0             1.0             0.0   
4                      0.0             1.0             0.0             1.0   
...                    ...             ...             ...             ...   
1196                   1.0             1.0             1.0             0.0   
1197                   0.0             1.0             1.0             0.0   
1198                   0.0             1.0             1.0             0.0   
1199                   1.0             1.0             0.0             1.0   
1200                   0.0             1.0             0.0             1.0   

      MechVent_med_0.0  MechVent_med_0.5  ihd  Label   Score  
0                  1.0               0.0    0      0  0.1476  
1                  1.0               0.0    0      0  0.4248  
2                  1.0               0.0    1      0  0.4440  
3                  1.0               0.0    0      0  0.0550  
4                  1.0               0.0    0      0  0.4130  
...                ...               ...  ...    ...     ...  
1196               1.0               0.0    1      1  0.5185  
1197               1.0               0.0    0      0  0.1041  
1198               1.0               0.0    1      0  0.2854  
1199               1.0               0.0    0      0  0.2974  
1200               1.0               0.0    0      0  0.2849  

[1201 rows x 125 columns]

## conclusion

similar to the last case of multicollinearity removal, building upon the imbalance-fixed model, through additional normalization / transformation, either a single model, e.g. gradient boosting classifier, or blend models has increased train or test min(recall, precision) to 0.44 or 0.42